pkg/api/authz.go

```diff
@@ -110,32 +130,78 @@ func (ac *AccessController) isAdmin(username string) bool {
 	return common.Contains(ac.Config.AdminPolicy.Users, username)
 }
 
+func (ac *AccessController) isAnyGroupInAdminPolicy(groups []string, username string) bool {
+	userGroups := ac.getUserGroups(groups, username)
+
+	for _, group := range userGroups {
+		if common.Contains(ac.Config.AdminPolicy.Groups, group) {
+			return true
+		}
+	}
+
+	return false
+}
+
 // getContext builds ac context(allowed to read repos and if user is admin) and returns it.
 func (ac *AccessController) getContext(username string, request *http.Request) context.Context {
-	readGlobPatterns := ac.getGlobPatterns(username, Read)
-	dmcGlobPatterns := ac.getGlobPatterns(username, DetectManifestCollision)
-
-	acCtx := localCtx.AccessControlContext{
-		ReadGlobPatterns: readGlobPatterns,
-		DmcGlobPatterns:  dmcGlobPatterns,
-		Username:         username,
+	acCtx, err := localCtx.GetAccessControlContext(request.Context())
+	if err != nil {
+		return nil
 	}
 
+	readGlobPatterns := ac.getGlobPatterns(username, acCtx.Groups, Read)
+	dmcGlobPatterns := ac.getGlobPatterns(username, acCtx.Groups, DetectManifestCollision)
+
+	acCtx.ReadGlobPatterns = readGlobPatterns
+	acCtx.DmcGlobPatterns = dmcGlobPatterns
+
 	if ac.isAdmin(username) {
 		acCtx.IsAdmin = true
 	} else {
 		acCtx.IsAdmin = false
 	}
 
 	authzCtxKey := localCtx.GetContextKey()
-	ctx := context.WithValue(request.Context(), authzCtxKey, acCtx)
+	ctx := context.WithValue(request.Context(), authzCtxKey, *acCtx)
 
 	return ctx
 }
 
+func (ac *AccessController) getUserGroups(groupNames []string, username string) []string {
+	for groupName, group := range ac.Config.Groups {
+		for _, user := range group.Users {
+			// find if the user is part of any groups
+			if user == username {
+				groupNames = append(groupNames, groupName)
+			}
+		}
+	}
+
+	return groupNames
+}
+
 // isPermitted returns true if username can do action on a repository policy.
-func isPermitted(username, action string, policyGroup config.PolicyGroup) bool {
+func (ac *AccessController) isPermitted(groups []string, username, action string,
+	policyGroup config.PolicyGroup,
+) bool {
 	var result bool
+
+	userGroups := ac.getUserGroups(groups, username)
```

Comment:
This shouldn't be here, the local groups should be set on the context by the the authentication handler as well.

Reply:
I moved the logic in authn.go